lib/reek/tree_walker.rb

```diff
@@ -18,24 +18,36 @@ module Reek
   #
   # :reek:TooManyMethods: { max_methods: 29 }
   class TreeWalker
-    def initialize(smell_repository, exp)
-      @smell_repository = smell_repository
-      @exp = exp
+    private_attr_accessor :element
+    private_attr_reader :exp, :context_tree
+
+    def initialize(syntax_tree)
+      @exp = syntax_tree
       @element = Context::RootContext.new(exp)
       @context_tree = process(exp)
     end
 
-    def walk
-      context_tree.each do |element|
-        smell_repository.examine(element)
+    # Walks the syntax tree, runs the configured SmellDetectors and returns
+    # a SmellRepository that contains the reported SmellWarnings.
+    #
+    # @param smell_types [Array<Reek::Smells::SmellDetector>] The smell types we want
+    #        to apply when walking the tree.
+    # @param configuration [Configuration::AppConfiguration] The corresponding configuration.
+
+    #
+    # @return [Reek::Smells::SmellRepository] SmellRepository that contains all used
+    #         SmellDetectors including all reported SmellWarnings.
+    def walk(smell_types: raise, configuration: raise)
+      Smells::SmellRepository.new(smell_types: smell_types,
+                                  configuration: configuration).tap do |smell_repository|
```

Comment:
Yes, I see your point about the magic workings.

Perhaps for now we can inline the entire `#walk` method into `Examiner#run`? I realize this sort of makes the name `TreeWalker` meaningless, but we can rename it to match the actual function it then thas, which is setting up the contexts.

This will then allow refactoring the `smell_repository.examine(...); smell_repository.report_on(...)` combination, which is another instance of the OOCP.

WDYT?